File: app/src/main/java/com/jsqix/dongqing/gank/fragment/douban/DoubanFilmFragment.java
package com.jsqix.dongqing.gank.fragment.douban;

import android.support.v7.widget.GridLayoutManager;
import android.util.Log;

import com.google.gson.Gson;
import com.jsqix.dongqing.gank.R;
import com.jsqix.dongqing.gank.adapter.recycler.DoubanFilmAdapter;
import com.jsqix.dongqing.gank.api.Api;
import com.jsqix.dongqing.gank.bean.douban.DoubanFilm;
import com.jsqix.dongqing.gank.fragment.RefreshFragment;
import com.lcodecore.tkrefreshlayout.RefreshListenerAdapter;
import com.lcodecore.tkrefreshlayout.TwinklingRefreshLayout;

import java.util.ArrayList;
import java.util.List;

import rx.Subscriber;
import rx.android.schedulers.AndroidSchedulers;
import rx.schedulers.Schedulers;

/**
 * Created by dongqing on 2017/1/12.
 */

public class DoubanFilmFragment extends RefreshFragment {
    private List<DoubanFilm.SubjectsBean> data = new ArrayList<>();
    private DoubanFilmAdapter adapter;

    @Override
    protected void initView() {
        refreshLayout.setEnableLoadmore(false);
        refreshLayout.setOverScrollBottomShow(false);
        GridLayoutManager manager = new GridLayoutManager(mContext, 2,GridLayoutManager.VERTICAL,false);
        recyclerView.setLayoutManager(manager);
        adapter = new DoubanFilmAdapter(mContext, R.layout.item_recycler_douban_film, data);
        recyclerView.setAdapter(adapter);

        refreshLayout.setOnRefreshListener(new RefreshListenerAdapter() {
            @Override
            public void onRefresh(TwinklingRefreshLayout refreshLayout) {
                getData();
            }
        });
    }

    private void getData() {
        if ("正在热映".equals(categoryType)) {
            getWell();
        } else if ("即将上线".equals(categoryType)) {
            getCome();
        }
    }

    private void getWell() {
        Api.getInstance().DOUBAN.service.getFilmWell()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new Subscriber<DoubanFilm>() {
                    @Override
                    public void onCompleted() {
                        Log.v("success:", "onCompleted");
                    }

                    @Override
                    public void onError(Throwable e) {
                        Log.v("failure:", "onError");
                        refreshLayout.finishRefreshing();
                    }

                    @Override
                    public void onNext(DoubanFilm doubanFilm) {
                        Log.v("next:", new Gson().toJson(doubanFilm));
                        refreshLayout.finishRefreshing();
                        data.clear();
                        data.addAll(doubanFilm.getSubjects());
                        adapter.notifyDataSetChanged();
                    }
                });
    }

    private void getCome() {
        Api.getInstance().DOUBAN.service.getFilmCome()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new Subscriber<DoubanFilm>() {
                    @Override
                    public void onCompleted() {
                        Log.v("success:", "onCompleted");
                    }

                    @Override
                    public void onError(Throwable e) {
                        Log.v("failure:", "onError");
                        refreshLayout.finishRefreshing();
                    }

                    @Override
                    public void onNext(DoubanFilm doubanFilm) {
                        Log.v("next:", new Gson().toJson(doubanFilm));
                        refreshLayout.finishRefreshing();
                        data.clear();
                        data.addAll(doubanFilm.getSubjects());
                        adapter.notifyDataSetChanged();
                    }
                });
    }
}

File: app/src/main/java/com/dongqing/gank/kotlin/ui/fragment/douban/DoubanFilmFragment.kt
package com.dongqing.gank.kotlin.ui.fragment.douban

import android.support.v7.widget.GridLayoutManager
import com.dongqing.gank.kotlin.adapter.DoubanFilmAdapter
import com.dongqing.gank.kotlin.bean.douban.DoubanFilm
import com.dongqing.gank.kotlin.ui.fragment.RefreshFragment
import com.dongqing.gank.kotlin.ui.present.DoubanPresent
import com.dongqing.gank.kotlin.ui.view.IDoubanView
import kotlinx.android.synthetic.main.fragment_base.*

class DoubanFilmFragment : RefreshFragment<DoubanPresent, IDoubanView>(), IDoubanView {

    private lateinit var adapter: DoubanFilmAdapter
    private var data = ArrayList<DoubanFilm.SubjectsBean>()

    override fun onDataReceive(t: DoubanFilm) {
        if (page === 1) {
            data.clear()
            if (refresh_layout != null) {
                refresh_layout.finishRefreshing()
            }
        } else {
            if (refresh_layout != null) {
                refresh_layout.finishLoadmore()
            }
        }
        if (t.subjects == null || t.subjects.size == 0) {
            toast("没有更多的数据了")
        } else {
            data.addAll(t.subjects)
            adapter.notifyDataSetChanged()
        }
    }

    override fun onDataError() {
        if (refresh_layout != null) {
            refresh_layout.finishRefreshing()
            refresh_layout.finishLoadmore()
        }
    }

    override fun initView() {
        super.initView()
        val manager = GridLayoutManager(mContext, 2, GridLayoutManager.VERTICAL, false)
        recyclerView.layoutManager = manager
        adapter = DoubanFilmAdapter(mContext, data)
        recyclerView.adapter = adapter
    }

    override fun getData() {
        if ("正在热映" == categoryType) {
            getWell()
        } else if ("即将上线" == categoryType) {
            getCome()
        }
    }

    private fun getWell() {
        mPresenter!!.getWellFilm(page)
    }

    private fun getCome() {
        mPresenter!!.getComeFilm(page)

    }
}